ethr/ethr.go

```diff
@@ -0,0 +1,473 @@
+func NewBoundContract(address common.Address, backend bind.ContractBackend) (*abi.ABI, *bind.BoundContract, error) {
+	parsed, err := abi.JSON(strings.NewReader(contracts.EthereumDIDRegistryABI))
+	if err != nil {
+		return nil, nil, err
+	}
+	return &parsed, bind.NewBoundContract(address, parsed, backend, backend, backend), nil
+}
+
+// InterpretIdentifier takes an input (hex) string and returns an address and/or public key string
+func InterpretIdentifier(identifier string) (common.Address, *ecdsa.PublicKey, error) {
+	if len(identifier) > 42 {
+		if strings.HasPrefix(identifier, "0x") {
+			identifier = identifier[2:]
+		}
+		publicKeyBytes, err := hex.DecodeString(identifier)
+		if err != nil {
+			return common.Address{}, nil, err
+		}
+		publicKey, err := crypto.DecompressPubkey(publicKeyBytes)
+		if err != nil {
+			return common.Address{}, nil, err
+		}
+		return crypto.PubkeyToAddress(*publicKey), publicKey, nil
+	}
+	return common.HexToAddress(identifier), nil, nil
+}
+
+func getLogs(client bind.ContractBackend, address common.Address, previousChange *big.Int) ([]*EthereumDIDRegistryDIDEventUnion, error) {
+	// Leading bytes are used here, because BytesToHash trims from the left
+	leading := common.Hex2Bytes("0x000000000000000000000000")
+	topic := common.BytesToHash(append(leading, address.Bytes()...))
+	query := ethereum.FilterQuery{
+		FromBlock: previousChange,
+		ToBlock:   previousChange,
+		Addresses: []common.Address{
+			ContractAddress,
+		},
+	}
+	// The first topic has to be nil to allow all event types through
+	// The second topic is set to the input address to get all events for a given
+	// address, otherwise, we'd get all updates for all addresses
+	query.Topics = append(query.Topics, nil, []common.Hash{topic})
+	// We use the lower-level FilterLogs here rather than the auto-built
+	// contract log filters because we want all the events, not just a single
+	// type.
+	logs, err := client.FilterLogs(context.Background(), query)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new filterer that is bound to the given address
+	abi, bound, err := NewBoundContract(address, client)
+	if err != nil {
+		return nil, err
+	}
+
+	// We keep all event types together as generic interface{} to enable in-order
+	// parsing later
+	events := make([]*EthereumDIDRegistryDIDEventUnion, len(logs))
+
+	// Parse the events in reverse order, so that we can build up the doc
+	for i := len(logs) - 1; i >= 0; i-- {
+		log := logs[i]
+		evt, err := abi.EventByID(log.Topics[0])
+		if err != nil {
+			return nil, err
+		}
+		e := new(EthereumDIDRegistryDIDEventUnion)
+		e.Type = evt.Name
+		if err = bound.UnpackLog(e, evt.Name, log); err != nil {
+			return nil, err
+		}
+		e.Raw = log
+		events[i] = e
+	}
+	return events, nil
+}
+
+func changeLog(client bind.ContractBackend, identifier string) (*common.Address, []*EthereumDIDRegistryDIDEventUnion, *ecdsa.PublicKey, error) {
```

Comment:
nit: `changeLog` sounds like we're changing the log... maybe that's happening here but not popping out to me